# Машинное обучение, MDS

## НИУ ВШЭ, 2025-26 учебный год

### Домашнее задание №1. Библиотеки Numpy & Pandas

Задание выполнил(а):

    Атаманюк Андрей

### Общая информация

__Дата выдачи:__ 01.02.2026

__Дедлайн:__ 04:00 13.02.2026

### Оценивание и штрафы

Каждая задача первого раздела (Numpy) оценивается в 2 балла.

Оценка за первый раздел вычисляется по следующей формуле:

$$
s_{\text{numpy}} \times 1/2 ,
$$

где $s_{\text{numpy}}$  — количество баллов, которое вы набрали в сумме по задачам первого раздела.

Каждая задача второго раздела (Pandas) оценивается в 3 балла.

Оценка за второй раздел вычисляется по следующей формуле:

$$
s_{\text{pandas}} \times 10/54 ,
$$

где $s_{\text{pandas}}$  — количество баллов, которое вы набрали в сумме по задачам второго раздела.

Общая оценка за ДЗ вычисляется по следующей формуле:

$$
O_{1} \times 3/10 + O_2 \times 7/10 ,
$$

где $O_1, O_2$  — оценки за первый и второй раздел соответственно.

За сдачу домашнего задания позже срока на итоговую оценку за задание накладывается штраф в размере 1 **вторичный** балл в день, но задержка не может быть больше недели.

**Важно:** В этом домашнем задании предусмотрено несколько заданий, за которые можно получить бонусные баллы. Бонусные баллы прибавляются к обычным баллам, и может выйти так, что ваша итоговая оценка станет больше 10. В таком случае именно это и будет являться вашей оценкой. То есть за это ДЗ **можно** получить больше 10.

__Внимание!__ Домашнее задание выполняется самостоятельно. Не допускается помощь в решении домашнего задания от однокурсников или третьих лиц. «Похожие» решения считаются плагиатом, и все задействованные студенты — в том числе и те, у кого списали, — не могут получить за него больше 0 баллов.

Использование в решении домашнего задания генеративных моделей (ChatGPT и так далее) допускается исключительно в рамках справочной и образовательной информации. Любые другие случаи применения средств ИИ — например, для автоматической генерации кода по заданию — считаются плагиатом, и такое домашнее задание оценивается в 0 баллов.

В случае если в решении используются подходы, функции или формулы, не разбиравшиеся на лекциях и семинарах, — необходимо обязательное указание источника: ссылка на документацию, статью, другой ресурс. При отсутствии оформления источников задание может быть полностью обнулено.

### Формат сдачи

Необходимо загружать файл с расширением .ipynb (питоновский ноутбук)

Ссылка для загрузки будет указана в нашем телеграмм-канале позднее.

## Раздел 1. Numpy

### О задании

В данном разделе вы попрактикуетесь в работе с библиотекой numpy!

In [1]:
import numpy as np

Во всех задачах необходимо написать код решения внутри функции и проверить его работу, вызвав функцию для данных из условия.

При решении задач запрещается использовать циклы (`for`, `while`) и оператор `if`.

Везде, где встречаются массивы или матрицы, подразумевается, что это `numpy.array`.

### 1

Напишите функцию, возвращающую округленную взвешенную сумму оценок по данным оценкам и весам.

In [2]:
def result_mark(weights, marks):
    return np.round(weights.dot(marks))

In [3]:
weights = np.array([0.25, 0.1, 0.2, 0.2, 0.1, 0.1, 0.05])
marks = np.array([6, 9, 8, 7, 10, 7, 10])
result_mark(weights, marks)

np.float64(8.0)

### 2

Напишите функцию, меняющую каждое третье (начиная с 0) значение массива целых чисел на заданное число. Например, если на вход поступает массив `array([3, 5, 1, 0, -3, 22, 213436])` и число `-111`, то на выходе должен получиться массив `array([-111, 5, 1, -111, -3, 22, -111])`.

In [4]:
def change_array(array, number):
    array[::3] = number
    return array

In [5]:
array = np.array([3, 5, 1, 0, -3, 22, 213436])
number = -111
change_array(array, number)

array([-111,    5,    1, -111,   -3,   22, -111])

### 3

Напишите функцию, выдающую индексы «близких» элементов заданных массивов, а именно тех пар элементов, чей модуль разницы не превосходит заданного значения. Например, если на вход поступают массив `array([1.5, 0.5, 2, -4.1, -3, 6, -1])`, массив `array([1.2, 0.5, 1, -4, 3, 0, -1.2])` и число `0.5`, то на выходе должен получиться массив `array([0, 1, 3, 6])` (_**важно: не `tuple`, а одномерный массив типа `numpy.ndarray` (то есть `.ndim` от него равно 1)!**_).

In [6]:
def find_close(array1, array2, precision):
    # https://numpy.org/devdocs/reference/generated/numpy.where.html - указываю только 
    # condition, поэтому ведёт себя как np.asarray(condition).nonzero() - т.е. возвращает 
    # индексы, где условие выполняется
    res = np.where(np.abs(array1 - array2) <= precision)
    return res[0] # первый элемент т.к. where возвращает tuple состоящий из numpy.ndarray

In [7]:
array1 = np.array([1.5, 0.5, 2, -4.1, -3, 6, -1])
array2 = np.array([1.2, 0.5, 1, -4, 3, 0, -1.2])
precision = 0.5
find_close(array1, array2, precision)

array([0, 1, 3, 6])

### 4

Напишите функцию, которая составляет блочную матрицу из четырех блоков, где каждый блок - это заданная матрица. Например, если на вход поступает матрица
$$
\begin{pmatrix}
0 & 1 & 2\\
3 & 4 & 5\\
\end{pmatrix},
$$
то ответом будет матрица
$$
\begin{pmatrix}
0 & 1 & 2 & 0 & 1 & 2\\
3 & 4 & 5 & 3 & 4 & 5\\
0 & 1 & 2 & 0 & 1 & 2\\
3 & 4 & 5 & 3 & 4 & 5\\
\end{pmatrix}
$$

In [8]:
def block_matrix(block):
    row = np.hstack((block, block))
    return np.vstack((row, row))

In [9]:
block = np.array([[0, 1, 2], [3, 4, 5]])
block_matrix(block)

array([[0, 1, 2, 0, 1, 2],
       [3, 4, 5, 3, 4, 5],
       [0, 1, 2, 0, 1, 2],
       [3, 4, 5, 3, 4, 5]])

### 5

Напишите функцию, вычисляющую произведение всех ненулевых диагональных элементов на главной диагонали данной квадратной матрицы. Например, если на вход поступает матрица
$$
\begin{pmatrix}
0 & 1 & 2\\
3 & 4 & 5\\
6 & 7 & 8\\
\end{pmatrix},
$$
то ответом будет 32.

In [10]:
def diag_prod(matrix):
    diag = np.diag(matrix) # https://numpy.org/devdocs/reference/generated/numpy.diag.html
    diag = diag[diag != 0]
    return np.prod(diag) # https://numpy.org/devdocs/reference/generated/numpy.prod.html

In [11]:
matrix = np.array([[0, 1, 2], [3, 4, 5], [6, 7, 8]])
diag_prod(matrix)

np.int64(32)

### 6

Для улучшения качества работы некоторых алгоритмов машинного обучения может быть полезно использовать [нормализацию данных](https://vk.cc/8xmfQk), чтобы привести признаки в выборке к одному масштабу — а именно, из каждого столбца вычесть среднее его значений и поделить на их стандартное отклонение. Напишите функцию, нормализующую входящую матрицу (по столбцам). Например, если на вход подается матрица
$$
\begin{pmatrix}
1 & 4 & 4200\\
0 & 10 & 5000\\
1 & 2 & 1000\\
\end{pmatrix},
$$
то результатом с точностью до сотых будет матрица
$$
\begin{pmatrix}
0.71 & -0.39 & 0.46\\
-1.41 & 1.37 & 0.93\\
0.71 & -0.98 & -1.39\\
\end{pmatrix}
$$

_Подсказка. Учтите, что в вашей матрице не должно получаться никаких nan. Подумайте, в каком случае они могут возникнуть и как обойти эту проблему._


In [12]:
def normalize(matrix):
    mean = matrix.sum(axis=0) / matrix.shape[0] # ну раз mean не проходили
    var = ((matrix - mean) ** 2).sum(axis=0) / matrix.shape[0]
    std = var ** 0.5

    std[std == 0] = 1 # избегаем деления на ноль

    matrix = (matrix - mean) / std
    return matrix

In [13]:
matrix = np.array([[1, 4, 4200], [0, 10, 5000], [1, 2, 1000]])
normalize(matrix)

array([[ 0.70710678, -0.39223227,  0.46291005],
       [-1.41421356,  1.37281295,  0.9258201 ],
       [ 0.70710678, -0.98058068, -1.38873015]])

### 7

Напишите функцию, вычисляющую какую-нибудь первообразную данного полинома (в качестве константы возьмите ваше любимое число). Например, если на вход поступает массив коэффициентов `array([4, 6, 0, 1])`, что соответствует полиному $4x^3 + 6x^2 + 1$, на выходе получается массив коэффициентов `array([1, 2, 0, 1, -2])`, соответствующий полиному $x^4 + 2x^3 + x - 2$.

In [14]:
def antiderivative(coefs):
    new_powers = np.arange(coefs.shape[0], 0, -1)
    return np.append(coefs / new_powers, 42)

In [15]:
coefs = np.array([4, 6, 0, 1])
antiderivative(coefs)

array([ 1.,  2.,  0.,  1., 42.])

### 8

Напишите функцию, делающую данную [треугольную матрицу](https://ru.wikipedia.org/wiki/%D0%A2%D1%80%D0%B5%D1%83%D0%B3%D0%BE%D0%BB%D1%8C%D0%BD%D0%B0%D1%8F_%D0%BC%D0%B0%D1%82%D1%80%D0%B8%D1%86%D0%B0) симметричной. Например, если на вход поступает матрица
$$
\begin{pmatrix}
1 & 2 & 3 & 4\\
0 & 5 & 6 & 7\\
0 & 0 & 8 & 9\\
0 & 0 & 0 & 10\\
\end{pmatrix},
$$
то на выходе должна быть матрица
$$
\begin{pmatrix}
1 & 2 & 3 & 4\\
2 & 5 & 6 & 7\\
3 & 6 & 8 & 9\\
4 & 7 & 9 & 10\\
\end{pmatrix}.
$$

In [16]:
def make_symmetric(matrix):
    transposed = matrix.T
    new = matrix + transposed
    # диагонально продублировалась
    new = new - matrix * np.identity(matrix.shape[0])
    return new

In [17]:
matrix = np.array([[1, 2, 3, 4], [0, 5, 6, 7], [0, 0, 8, 9], [0, 0, 0, 10]])
make_symmetric(matrix)

array([[ 1.,  2.,  3.,  4.],
       [ 2.,  5.,  6.,  7.],
       [ 3.,  6.,  8.,  9.],
       [ 4.,  7.,  9., 10.]])

### 9

Напишите функцию, создающую прямоугольную матрицу из m одинаковых строк, заполненных последовательными натуральными числами от a до b включительно в возрастающем порядке. Например, если m = 5, a = 3, b = 10, то на выходе будет матрица
$$
\begin{pmatrix}
3 & 4 & 5 & 6 & 7 & 8 & 9 & 10\\
3 & 4 & 5 & 6 & 7 & 8 & 9 & 10\\
3 & 4 & 5 & 6 & 7 & 8 & 9 & 10\\
3 & 4 & 5 & 6 & 7 & 8 & 9 & 10\\
3 & 4 & 5 & 6 & 7 & 8 & 9 & 10\\
\end{pmatrix}
$$

In [18]:
def construct_matrix(m, a, b):
    row = np.arange(a, b + 1)
    filler_matrix = np.ones((m, row.shape[0]))
    return filler_matrix * row

In [19]:
m = 5
a = 3
b = 10
construct_matrix(m, a, b)

array([[ 3.,  4.,  5.,  6.,  7.,  8.,  9., 10.],
       [ 3.,  4.,  5.,  6.,  7.,  8.,  9., 10.],
       [ 3.,  4.,  5.,  6.,  7.,  8.,  9., 10.],
       [ 3.,  4.,  5.,  6.,  7.,  8.,  9., 10.],
       [ 3.,  4.,  5.,  6.,  7.,  8.,  9., 10.]])

### 10

Напишите функцию, вычисляющую [косинусную близость](https://en.wikipedia.org/wiki/Cosine_similarity) двух векторов. Например, если на вход поступают вектора `array([-2, 1, 0, -5, 4, 3, -3])` и `array([0, 2, -2, 10, 6, 0, 0])`, ответом будет -0.25.

In [20]:
def cosine_similarity(vec1, vec2):
    dot = vec1.dot(vec2)

    norm1 = (vec1 ** 2).sum() ** 0.5
    norm2 = (vec2 ** 2).sum() ** 0.5

    return dot / (norm1 * norm2)

In [21]:
vec1 = np.array([-2, 1, 0, -5, 4, 3, -3])
vec2 = np.array([0, 2, -2, 10, 6, 0, 0])
cosine_similarity(vec1, vec2)

np.float64(-0.25)

## Раздел 2. Pandas

### О задании

В данном разделе вы попрактикуетесь в работе с библиотекой pandas!

In [22]:
import pandas as pd

Обратите внимание, что практически во всех заданиях вам потребуется написать какой-то текст в виде вывода/результата/итога/какого-то резюме. Пожалуйста, не ленитесь, и пишите выводы ***подробно***, существенная часть баллов будет выставляться именно за выводы, так как в анализе данных они имеют большое значение. Таким образом, на любое задание, сформулированное в виде вопроса, отвечайте текстом, подкрепляя ваши слова написанным и выполненным кодом. Текст выводов/комментариев пишите в отдельных ячейках, используя средства текстового редактора Jupyter Notebook/Google Colab, при необходимости используйте средства *выделения*.

При решении задач, когда вам необходимо сделать какой-то фильтр, срез, выборку, или иное действие с данными, запрещается использовать циклы (`for`, `while`) и оператор `if`.


### 1

В этом разделе домашнего задания вы поработаете с данными исследования зависимости средней продолжительности жизни по странам от количества употребляемого в стране алкоголя.

Вашему вниманию представлено два датасета: drinks.csv и lifeexpectancy-verbose.csv

Первая таблица содержит в себе сводные данные по количеству употребляемого алкоголя в каждой стране.

Описание датасета:
- country -	Name of the country (String)
- beer_servings	- Number of beer servings per person in the country (Integer)
- spirit_servings	- Number of spirit servings per person in the country (Integer)
- wine_servings	- Number of wine servings per person in the country(Integer)
- total_litres_of_pure_alcohol -	Total litres of pure alcohol consumed per person in the country (Float)

Вторая таблица содержит в себе информацию о научных статьях и исследованиях о продолжительности жизни (в разных категориях и для разных групп) по той или иной стране.

Откройте в ноутбуке оба файла с данными. Какой размер имеют таблицы? Выведите первые 5 и последние 5 строк для обеих таблиц. Сколько уникальных стран представлено в каждой из таблиц? Сделайте выводы.

In [23]:
df1 = pd.read_csv('drinks.csv')

print(df1.shape[0])
df1.head()

193


index      country  beer_servings  spirit_servings  wine_servings  \
0      0  Afghanistan              0                0              0   
1      1      Albania             89              132             54   
2      2      Algeria             25                0             14   
3      3      Andorra            245              138            312   
4      4       Angola            217               57             45   

   total_litres_of_pure_alcohol  
0                           0.0  
1                           4.9  
2                           0.7  
3                          12.4  
4                           5.9

In [24]:
df1.tail()

index    country  beer_servings  spirit_servings  wine_servings  \
188    188  Venezuela            333              100              3   
189    189    Vietnam            111                2              1   
190    190      Yemen              6                0              0   
191    191     Zambia             32               19              4   
192    192   Zimbabwe             64               18              4   

     total_litres_of_pure_alcohol  
188                           7.7  
189                           2.0  
190                           0.1  
191                           2.5  
192                           4.7

In [25]:
print(df1.country.nunique())

193


In [26]:
df2 = pd.read_csv("lifeexpectancy-verbose.csv")

print(df2.shape[0])
df2.head()

6408


index        GhoCode                                       GhoDisplay  \
0      0  WHOSIS_000001                 Life expectancy at birth (years)   
1      1  WHOSIS_000002  Healthy life expectancy (HALE) at birth (years)   
2      2  WHOSIS_000001                 Life expectancy at birth (years)   
3      3  WHOSIS_000001                 Life expectancy at birth (years)   
4      4  WHOSIS_000015                Life expectancy at age 60 (years)   

  PublishStateCode PublishStateDisplay  YearCode  YearDisplay RegionCode  \
0        PUBLISHED           Published      1990         1990        AMR   
1        PUBLISHED           Published      2012         2012        EUR   
2        PUBLISHED           Published      2000         2000        EMR   
3        PUBLISHED           Published      2000         2000        AFR   
4        PUBLISHED           Published      1990         1990       SEAR   

           RegionDisplay WorldBankIncomeGroupGroupCode  \
0               Americas                        WB_LMI   
1                 Europe                         WB_HI   
2  Eastern Mediterranean                         WB_LI   
3                 Africa                        WB_LMI   
4        South_East Asia                        WB_LMI   

  WorldBankIncomeGroupDisplay CountryCode CountryDisplay SexCode  SexDisplay  \
0         Lower_middle_income         NIC      Nicaragua     MLE        Male   
1                 High_income         IRL        Ireland     MLE        Male   
2                  Low_income         YEM          Yemen    BTSX  Both sexes   
3         Lower_middle_income         NGA        Nigeria    FMLE      Female   
4         Lower_middle_income         THA       Thailand    FMLE      Female   

   DisplayValue  Numeric  
0            68     68.0  
1            69     69.3  
2            61     61.0  
3            48     48.0  
4            19     19.0

In [27]:
df2.tail()

index        GhoCode                                       GhoDisplay  \
6403   6403  WHOSIS_000001                 Life expectancy at birth (years)   
6404   6404  WHOSIS_000015                Life expectancy at age 60 (years)   
6405   6405  WHOSIS_000001                 Life expectancy at birth (years)   
6406   6406  WHOSIS_000002  Healthy life expectancy (HALE) at birth (years)   
6407   6407  WHOSIS_000001                 Life expectancy at birth (years)   

     PublishStateCode PublishStateDisplay  YearCode  YearDisplay RegionCode  \
6403        PUBLISHED           Published      2000         2000        EMR   
6404        PUBLISHED           Published      1990         1990        WPR   
6405        PUBLISHED           Published      1990         1990        AFR   
6406        PUBLISHED           Published      2012         2012        EUR   
6407        PUBLISHED           Published      2013         2013        EUR   

              RegionDisplay WorldBankIncomeGroupGroupCode  \
6403  Eastern Mediterranean                        WB_LMI   
6404        Western Pacific                        WB_UMI   
6405                 Africa                         WB_LI   
6406                 Europe                        WB_LMI   
6407                 Europe                           NaN   

     WorldBankIncomeGroupDisplay CountryCode CountryDisplay SexCode  \
6403         Lower_middle_income         MAR        Morocco    BTSX   
6404         Upper_middle_income         PLW          Palau    BTSX   
6405                  Low_income         BFA   Burkina Faso    FMLE   
6406         Lower_middle_income         ALB        Albania     MLE   
6407                         NaN         ARM        Armenia    BTSX   

      SexDisplay  DisplayValue   Numeric  
6403  Both sexes            68  68.00000  
6404  Both sexes            16  16.00000  
6405      Female            51  51.00000  
6406        Male            64  63.50000  
6407  Both sexes            71  71.02699

In [28]:
print(df2.CountryDisplay.nunique())

194


#### Выводы:

пока получается сделать вывод только структуре таблиц: гранулярность второй гораздо глубже - первая содержит информацию по странам без дублей, а вторая содержит для каждой страны ещё несколько строк с информацией по разным значениям других колонок.

P.S. не особо понятен смысл данных в первой таблице - это сводка за какой промежуток времени? За всю историю? За последний год? Какие порции алкоголя?

### 2

Для каждого признака в рассматриваемых данных укажите вид этого признака и кратко обоснуйте, почему вид именно такой.

In [29]:
df1.dtypes

index                             int64
country                             str
beer_servings                     int64
spirit_servings                   int64
wine_servings                     int64
total_litres_of_pure_alcohol    float64
dtype: object

In [30]:
df1.iloc[0, :]

index                                     0
country                         Afghanistan
beer_servings                             0
spirit_servings                           0
wine_servings                             0
total_litres_of_pure_alcohol            0.0
Name: 0, dtype: object

In [31]:
df2.dtypes

index                              int64
GhoCode                              str
GhoDisplay                           str
PublishStateCode                     str
PublishStateDisplay                  str
YearCode                           int64
YearDisplay                        int64
RegionCode                           str
RegionDisplay                        str
WorldBankIncomeGroupGroupCode        str
WorldBankIncomeGroupDisplay          str
CountryCode                          str
CountryDisplay                       str
SexCode                              str
SexDisplay                           str
DisplayValue                       int64
Numeric                          float64
dtype: object

In [32]:
df2.iloc[0, :]

index                                                           0
GhoCode                                             WHOSIS_000001
GhoDisplay                       Life expectancy at birth (years)
PublishStateCode                                        PUBLISHED
PublishStateDisplay                                     Published
YearCode                                                     1990
YearDisplay                                                  1990
RegionCode                                                    AMR
RegionDisplay                                            Americas
WorldBankIncomeGroupGroupCode                              WB_LMI
WorldBankIncomeGroupDisplay                   Lower_middle_income
CountryCode                                                   NIC
CountryDisplay                                          Nicaragua
SexCode                                                       MLE
SexDisplay                                                   Male
DisplayVal

#### Вывод:

в первой таблице:

country — номинальный (просто названия, без порядка)

остальные три про servings — дискретные, так как это штуки (порции), а pure_alcohol — непрерывный, это литры, может быть дробным

во второй таблице:

всякие коды и display (country, region, sex) — номинальные, просто категории

year — дискретный, (или порядковый), это даты

numeric — это сама продолжительность жизни, непрерывный (там есть дробные значения)

world bank group — порядковый, так как есть иерархия доходов (low -> high)

### 3

Внимательно изучите признаки обеих таблиц. Как вы считаете, какая из таблиц является главной (а какая сопутствующей)? Почему вы так думаете?

Какой признак может представлять существенный интерес для исследования и являться целевой переменной в рамках представленного датасета? Ответ обоснуйте.

***Бонус (1 бон. балл):*** Попробуйте описать задачу машинного обучения, которую вы могли бы решать для данного датасета с учетом выбранной вами раннее целевой переменной и проблематики исследования. Подробнее про постановку и формализацию задач машинного обучения, а также про различные типы таких задач — вы можете прочитать, например, вот в этой [статье](https://habr.com/ru/company/simbirsoft/blog/565396/) (в начале)

#### Вывод:

главная таблица тут lifeexpectancy (которая df2). там данных намного больше (6к+ строк), она детальная - разбивка по годам, полу, расе и остальному.

а таблица drinks (df1) содержит меньше информации и выглядит как что-то, что нужно джоинить к более детальной.

**Бонус:** нам может быть интересно **объяснять** (не предсказывать для какого-то конкретного случая, а именно интерпретировать картину целиком), от каких параметров и в какой степени зависит продолжительность жизни. Задача - интерпретация коэффициентов модели линейной регрессии. Если проще формулировать - задача линейной регрессии.


### 4

Какова доля статей о продолжительности жизни в Африке? Какова доля статей о продолжительности жизни мужчин и доля статей о продолжительности жизни женщин среди всех статей по Африке?

***Бонус (2 бон. балла):*** Сравните эти доли с аналогичными долями по другим регионам. Как вы можете объяснить возникающие сходства и различия? Прокомментируйте полученные результаты.

In [33]:
africa_share = (df2['RegionDisplay'] == 'Africa').mean()
print(f'доля статей про Африку: {africa_share}')

доля статей про Африку: 0.23782771535580524


In [34]:
africa = df2[df2['RegionDisplay'] == 'Africa']

print('доли по полу в Африке:')
print(africa['SexDisplay'].value_counts(normalize=True))

доли по полу в Африке:
SexDisplay
Female        0.333333
Both sexes    0.333333
Male          0.333333
Name: proportion, dtype: float64


In [35]:
df2.groupby('RegionDisplay')['SexDisplay'].value_counts(normalize=True).unstack()

SexDisplay             Both sexes    Female      Male
RegionDisplay                                        
Africa                   0.333333  0.333333  0.333333
Americas                 0.333333  0.333333  0.333333
Eastern Mediterranean    0.333333  0.333333  0.333333
Europe                   0.333333  0.333333  0.333333
South_East Asia          0.333333  0.333333  0.333333
Western Pacific          0.333333  0.333333  0.333333

#### Вывод:

доля статей про Африку - почти четверть от всего датасета. Внутри Африки распределение статей по полам равномерное - более того, если смотреть такое же распределенеи статей по полу в других странах - всё тоже равномерно. Общий вывод про то, как собирался этот датасет - данные собирались равномерно.

### 5

Сколько всего существует видов статей с точки зрения оцениваемого параметра продолжительности жизни? Какой из этих видов статей самый популярный (написано наибольшее количество научных работ)? В рамках найденного вами самого популярного вида статей проведите сравнение, относительно какой из половых групп: мужчины, женщины или же оба пола — статей в нем было больше всего?

In [36]:
# видов статей:
df2['GhoDisplay'].unique()

<StringArray>
[               'Life expectancy at birth (years)',
 'Healthy life expectancy (HALE) at birth (years)',
               'Life expectancy at age 60 (years)']
Length: 3, dtype: str

In [37]:
# распределение частотности по типам
df2['GhoDisplay'].value_counts()

GhoDisplay
Life expectancy at birth (years)                   2328
Life expectancy at age 60 (years)                  2328
Healthy life expectancy (HALE) at birth (years)    1752
Name: count, dtype: int64

т.е. всего 3 типа статей, самых популярных два - Life expectancy at birth (years) и Life expectancy at age 60 (years)

In [38]:
top_type_1 = "Life expectancy at birth (years)"

df2[df2.GhoDisplay == top_type_1]["SexDisplay"].value_counts()

SexDisplay
Male          776
Both sexes    776
Female        776
Name: count, dtype: int64

In [39]:
top_type_2 = "Life expectancy at age 60 (years)"

df2[df2.GhoDisplay == top_type_2]["SexDisplay"].value_counts()

SexDisplay
Female        776
Both sexes    776
Male          776
Name: count, dtype: int64

#### Вывод:

из 3 типов первенство делят "Life expectancy at birth (years)" и "Life expectancy at age 60 (years)". По полу всё равномерно - данные собирались равномерно (ради сбалансированности пропорций выборок? Внутри регионов всё равно по полу и внутри типов статей всё равно по полу)

### 6

Выведите минимальную, среднюю, медианную и максимальную оцененную продолжительность жизни людей от рождения (life expectancy at birth). Выберите любые три страны на ваш вкус и проделайте то же самое только для статей по этим странам. Что можно сказать на основе результатов, которые вы получили?

In [40]:
type = "Life expectancy at birth (years)"

birth = df2.loc[df2["GhoDisplay"] == type, ["CountryDisplay", "Numeric"]]

birth

CountryDisplay   Numeric
0         Nicaragua  68.00000
2             Yemen  61.00000
3           Nigeria  48.00000
5             Chile  80.00000
6          Zimbabwe  58.72677
...             ...       ...
6396         Sweden  75.00000
6401       Honduras  69.00000
6403        Morocco  68.00000
6405   Burkina Faso  51.00000
6407        Armenia  71.02699

[2328 rows x 2 columns]

In [41]:
birth["Numeric"].agg(["min", "mean", "median", "max"])

min       38.000000
mean      68.441379
median    70.857895
max       87.154570
Name: Numeric, dtype: float64

In [42]:
# [x for x in birth.CountryDisplay.unique() if x.startswith("Chi")]
# цикл - не часть решения, просто интересно было найти правильное название

In [43]:
countries = ["Russian Federation", "Nigeria", "China"]

birth.loc[birth["CountryDisplay"].isin(countries)]\
    .groupby("CountryDisplay")["Numeric"]\
    .agg(["min", "mean", "median", "max"])

min       mean  median       max
CountryDisplay                                       
China               67.0  72.797656    73.5  77.00000
Nigeria             45.0  50.473338    50.5  55.40039
Russian Federation  59.0  68.078784    69.0  75.22597

#### Выводы:

в общем срезе медиана выше среднего (70 > 68) - значит, есть левый (в сторону нуля) хвост. Перекос в сторону низких значений (логично: возможно, люди часто умирают в молодости по разным причинам, а до поздних лет мало кто доживает - учитывая max меньше 100, всякие супер-аномальные исследования особых групп людей исключены).

Нигерия хуже России и Китая по всем этим метрикам.

Важно: эти выводы - это аггрегация по всем годам и по всем полам вместе.

### 7

Сформируйте результирующий датафрейм на основе проделанных вами действий в предыдущем номере. У датафрейма должно быть 4 столбца, в соответствии с заданными агреггирующими функциями: минимум, среднее, медиана, максимум, — а также 4 строки: статистика по всему датасету; статистика по трем выбранным вами странам. Сохраните программно созданный вами датафрейм в виде csv файла.

*Примечание. Датафрейм необходимо создавать автоматизированным образом. Не допускается ручной перенос значений из полученных в предыдущем задании ячеек в пустой датафрейм и/или прописывание конкретных значений (чисел) в коде программы. Представьте, что у нас не 3 страны, а 100 и не 4 агреггирующих метрики, а 100!*

In [44]:
type = "Life expectancy at birth (years)"
countries = ["Russian Federation", "Nigeria", "China"]

birth = df2.loc[df2["GhoDisplay"] == type, ["CountryDisplay", "Numeric"]]

stats_all = birth["Numeric"].agg(["min", "mean", "median", "max"])
stats_all_df = pd.DataFrame([stats_all], index=["All"])

stats_countries = (
    birth.loc[birth["CountryDisplay"].isin(countries)]
        .groupby("CountryDisplay")["Numeric"]
        .agg(["min", "mean", "median", "max"])
)

result = pd.concat([stats_all_df, stats_countries])

result.to_csv("task_2_7_result.csv")
result

min       mean     median       max
All                 38.0  68.441379  70.857895  87.15457
China               67.0  72.797656  73.500000  77.00000
Nigeria             45.0  50.473338  50.500000  55.40039
Russian Federation  59.0  68.078784  69.000000  75.22597

### 8

Есть ли в наших датасетах пропуски? Если да, то в каких столбцах? Сколько пропусков в каждом из них?

Подумайте, как стоит избавиться от пропусков в каждом из столбцов, и сделайте это. Аргументируйте ваши действия.

Проверьте, что пропусков нигде не осталось.

*Примечание. Не забудьте сохранить все изменения! В последующих заданиях вы должны работать с уже обновленными датасетами.*

In [45]:
df1.loc[df1.isna().sum(axis=1) > 0]

Empty DataFrame
Columns: [index, country, beer_servings, spirit_servings, wine_servings, total_litres_of_pure_alcohol]
Index: []

In [46]:
df2.loc[df2.isna().sum(axis=1) > 0]

index        GhoCode                         GhoDisplay  \
6         6  WHOSIS_000001   Life expectancy at birth (years)   
7         7  WHOSIS_000015  Life expectancy at age 60 (years)   
8         8  WHOSIS_000001   Life expectancy at birth (years)   
17       17  WHOSIS_000001   Life expectancy at birth (years)   
19       19  WHOSIS_000015  Life expectancy at age 60 (years)   
...     ...            ...                                ...   
6388   6388  WHOSIS_000001   Life expectancy at birth (years)   
6389   6389  WHOSIS_000001   Life expectancy at birth (years)   
6391   6391  WHOSIS_000015  Life expectancy at age 60 (years)   
6402   6402  WHOSIS_000015  Life expectancy at age 60 (years)   
6407   6407  WHOSIS_000001   Life expectancy at birth (years)   

     PublishStateCode PublishStateDisplay  YearCode  YearDisplay RegionCode  \
6           PUBLISHED           Published      2013         2013        AFR   
7           PUBLISHED           Published      2013         2013        EUR   
8           PUBLISHED           Published      2013         2013        EUR   
17          PUBLISHED           Published      2013         2013        AMR   
19          PUBLISHED           Published      2013         2013        WPR   
...               ...                 ...       ...          ...        ...   
6388        PUBLISHED           Published      2013         2013        EUR   
6389        PUBLISHED           Published      2013         2013        EUR   
6391        PUBLISHED           Published      2013         2013        WPR   
6402        PUBLISHED           Published      2013         2013       SEAR   
6407        PUBLISHED           Published      2013         2013        EUR   

        RegionDisplay WorldBankIncomeGroupGroupCode  \
6              Africa                           NaN   
7              Europe                           NaN   
8              Europe                           NaN   
17           Americas                           NaN   
19    Western Pacific                           NaN   
...               ...                           ...   
6388           Europe                           NaN   
6389           Europe                           NaN   
6391  Western Pacific                           NaN   
6402  South_East Asia                           NaN   
6407           Europe                           NaN   

     WorldBankIncomeGroupDisplay CountryCode CountryDisplay SexCode  \
6                            NaN         ZWE       Zimbabwe    BTSX   
7                            NaN         DEU        Germany    BTSX   
8                            NaN         ESP          Spain     MLE   
17                           NaN         BRA         Brazil     MLE   
19                           NaN         NIU           Niue    FMLE   
...                          ...         ...            ...     ...   
6388                         NaN         BEL        Belgium    BTSX   
6389                         NaN         SRB         Serbia    FMLE   
6391                         NaN         COK   Cook Islands     MLE   
6402                         NaN         TLS    Timor_Leste     MLE   
6407                         NaN         ARM        Armenia    BTSX   

      SexDisplay  DisplayValue   Numeric  
6     Both sexes            59  58.72677  
7     Both sexes            24  23.76277  
8           Male            80  79.56665  
17          Male            72  71.61015  
19        Female            21  20.83867  
...          ...           ...       ...  
6388  Both sexes            80  80.48828  
6389      Female            77  77.36535  
6391        Male            20  19.71973  
6402        Male            16  16.03829  
6407  Both sexes            71  71.02699  

[1164 rows x 17 columns]

In [47]:
df2.columns

Index(['index', 'GhoCode', 'GhoDisplay', 'PublishStateCode',
       'PublishStateDisplay', 'YearCode', 'YearDisplay', 'RegionCode',
       'RegionDisplay', 'WorldBankIncomeGroupGroupCode',
       'WorldBankIncomeGroupDisplay', 'CountryCode', 'CountryDisplay',
       'SexCode', 'SexDisplay', 'DisplayValue', 'Numeric'],
      dtype='str')

In [48]:
df2.isna().sum(axis=0) > 0

index                            False
GhoCode                          False
GhoDisplay                       False
PublishStateCode                 False
PublishStateDisplay              False
YearCode                         False
YearDisplay                      False
RegionCode                       False
RegionDisplay                    False
WorldBankIncomeGroupGroupCode     True
WorldBankIncomeGroupDisplay       True
CountryCode                      False
CountryDisplay                   False
SexCode                          False
SexDisplay                       False
DisplayValue                     False
Numeric                          False
dtype: bool

пропуски присутствуют только в WorldBankIncomeGroupGroupCode и WorldBankIncomeGroupDisplay       

In [49]:
df2.columns[df2.isna().sum(axis=0) > 0]

Index(['WorldBankIncomeGroupGroupCode', 'WorldBankIncomeGroupDisplay'], dtype='str')

In [50]:
len(df2[df2.WorldBankIncomeGroupGroupCode.isna()]), len(df2[df2.WorldBankIncomeGroupDisplay.isna()])

(1164, 1164)

если избавляться от колонок без мл (без предсказаний), то можно либо заполнить каким-то из имеющихся значений (например, самым частотным), либо заполнить категорией "unknown". 

Я бы не выбирал первый вариант: дело в том, что заполнять самым частотным - это тоже предсказание (забыли проставить категорию? ну, **вероятнее всего** это была категория, имеющая наибольшую вероятность - а вероятность мы оцениваем эмпирически по датасету). 

Но вряд ли здесь "забыли" что-то проставить. Может быть, что факт пропуска категории обоснован другими факторами - мы могли бы подробнее изучить и попробовать найти зависимость, например, что во всех недавних годах YearCode (например, после 2013) систематически этот фактор отсутствует. Каким бы ни был результат исследования, информативнее для дальнейшего анализа не "удалять" инфу о "пропущенности", затирая её most frequent значением. Лучше выставить новую категорию "unknown".

In [51]:
df2["WorldBankIncomeGroupGroupCode"] = df2["WorldBankIncomeGroupGroupCode"].fillna("unknown")
df2["WorldBankIncomeGroupDisplay"] = df2["WorldBankIncomeGroupDisplay"].fillna("unknown")

In [52]:
df2.loc[df2.isna().sum(axis=1) > 0]

Empty DataFrame
Columns: [index, GhoCode, GhoDisplay, PublishStateCode, PublishStateDisplay, YearCode, YearDisplay, RegionCode, RegionDisplay, WorldBankIncomeGroupGroupCode, WorldBankIncomeGroupDisplay, CountryCode, CountryDisplay, SexCode, SexDisplay, DisplayValue, Numeric]
Index: []

### 9

Все ли признаки несут в себе полезную информацию? Почему вы так считаете? Избавьтесь от ненужных столбцов. Свой выбор аргументируйте

*Примечание. Не забудьте сохранить все изменения! В последующих заданиях вы должны работать с уже обновленными датасетами.*

In [53]:
len(df1[df1["index"] != df1.index]), len(df2[df2["index"] != df2.index])

(0, 0)

значит, колонка index дублирует индекс таблицы - колонка лишняя.

In [54]:
df1 = df1.drop(columns=["index"])
df2 = df2.drop(columns=["index"])

больше в df1 ничего не удаляю - там колонки по смыслу про разную инфу

теперь - про DisplayValue и Numeric: методом пристального всматривания я обнаружил странные прецеденты округления:

In [55]:
# 5941, 6366
df2.loc[5941, ["DisplayValue", "Numeric"]]

DisplayValue    63.0
Numeric         63.5
Name: 5941, dtype: float64

In [56]:
df2.loc[6366, ["DisplayValue", "Numeric"]]

DisplayValue    59.0
Numeric         58.5
Name: 6366, dtype: float64

Я пока не могу быть уверенным, следует ли эта разница в стороне округления из каких-то скрытых причин или нет. Просто оставлю обе - DisplayValue и Numeric.

In [57]:
df2[["YearCode", "YearDisplay"]].groupby("YearCode")["YearDisplay"].value_counts()

YearCode  YearDisplay
1990      1990           1164
2000      2000           1749
2012      2012           1749
2013      2013           1746
Name: count, dtype: int64

соответствие 1-в-1

In [58]:
df2[["GhoCode", "GhoDisplay"]].groupby("GhoCode")["GhoDisplay"].value_counts()

GhoCode        GhoDisplay                                     
WHOSIS_000001  Life expectancy at birth (years)                   2328
WHOSIS_000002  Healthy life expectancy (HALE) at birth (years)    1752
WHOSIS_000015  Life expectancy at age 60 (years)                  2328
Name: count, dtype: int64

соответствие 1-в-1

In [59]:
df2[["CountryCode", "CountryDisplay"]].groupby("CountryCode")["CountryDisplay"].value_counts()

CountryCode  CountryDisplay      
AFG          Afghanistan             33
AGO          Angola                  33
ALB          Albania                 33
AND          Andorra                 33
ARE          United Arab Emirates    33
                                     ..
WSM          Samoa                   33
YEM          Yemen                   33
ZAF          South Africa            33
ZMB          Zambia                  33
ZWE          Zimbabwe                33
Name: count, Length: 194, dtype: int64

сходу непонятно. Попробуем иначе, с использованием apply + lambda (к каждой сущности .groupby("CountryCode")["CountryDisplay"] применяем lambda-функцию, которая посчитает ДЛИНУ ТАБЛИЦЫ value_counts). Если где-то больше 1 - значит, какой-то CountryCode встречается с более чем одним CountryDisplay:

In [60]:
(
    df2.groupby("CountryCode")["CountryDisplay"].apply(lambda x: x.value_counts().shape[0]) 
    != 1
).sum()

np.int64(0)

супер, соответствие 1-в-1

In [61]:
df2[["PublishStateCode", "PublishStateDisplay"]].groupby("PublishStateCode")["PublishStateDisplay"].value_counts()

PublishStateCode  PublishStateDisplay
PUBLISHED         Published              6408
Name: count, dtype: int64

бессмысленные колонки с константным значением. Обе вырезаем

In [62]:
(
    df2.groupby("RegionCode")["RegionDisplay"].apply(lambda x: x.value_counts().shape[0]) 
    != 1
).sum()

np.int64(0)

1-в-1

In [63]:
(
    df2.groupby("WorldBankIncomeGroupGroupCode")["WorldBankIncomeGroupDisplay"].apply(lambda x: x.value_counts().shape[0]) 
    != 1
).sum()

np.int64(0)

1-в-1

In [64]:
df2[["SexCode", "SexDisplay"]].groupby("SexCode")["SexDisplay"].value_counts()

SexCode  SexDisplay
BTSX     Both sexes    2136
FMLE     Female        2136
MLE      Male          2136
Name: count, dtype: int64

1-в-1

#### Вывод: удаляем все дублирующиеся code

In [65]:
df2 = df2[["GhoDisplay", "YearDisplay", "RegionDisplay", "WorldBankIncomeGroupDisplay", "CountryDisplay", "SexDisplay", "DisplayValue", "Numeric"]]

df2

GhoDisplay  YearDisplay  \
0                    Life expectancy at birth (years)         1990   
1     Healthy life expectancy (HALE) at birth (years)         2012   
2                    Life expectancy at birth (years)         2000   
3                    Life expectancy at birth (years)         2000   
4                   Life expectancy at age 60 (years)         1990   
...                                               ...          ...   
6403                 Life expectancy at birth (years)         2000   
6404                Life expectancy at age 60 (years)         1990   
6405                 Life expectancy at birth (years)         1990   
6406  Healthy life expectancy (HALE) at birth (years)         2012   
6407                 Life expectancy at birth (years)         2013   

              RegionDisplay WorldBankIncomeGroupDisplay CountryDisplay  \
0                  Americas         Lower_middle_income      Nicaragua   
1                    Europe                 High_income        Ireland   
2     Eastern Mediterranean                  Low_income          Yemen   
3                    Africa         Lower_middle_income        Nigeria   
4           South_East Asia         Lower_middle_income       Thailand   
...                     ...                         ...            ...   
6403  Eastern Mediterranean         Lower_middle_income        Morocco   
6404        Western Pacific         Upper_middle_income          Palau   
6405                 Africa                  Low_income   Burkina Faso   
6406                 Europe         Lower_middle_income        Albania   
6407                 Europe                     unknown        Armenia   

      SexDisplay  DisplayValue   Numeric  
0           Male            68  68.00000  
1           Male            69  69.30000  
2     Both sexes            61  61.00000  
3         Female            48  48.00000  
4         Female            19  19.00000  
...          ...           ...       ...  
6403  Both sexes            68  68.00000  
6404  Both sexes            16  16.00000  
6405      Female            51  51.00000  
6406        Male            64  63.50000  
6407  Both sexes            71  71.02699  

[6408 rows x 8 columns]

### 10

Посчитайте, насколько сильно коррелируют друг с другом потребления разных видов алкоголя (beer_servings,	spirit_servings и	wine_servings). Создайте датафрейм с попарными корреляциями этих столбцов друг с другом. Какие выводы можно сделать?

In [78]:
df1[["beer_servings", "spirit_servings", "wine_servings"]].corr()

beer_servings  spirit_servings  wine_servings
beer_servings         1.000000         0.458819       0.527172
spirit_servings       0.458819         1.000000       0.194797
wine_servings         0.527172         0.194797       1.000000

#### Выводы: потребления всех видов алкоголя коррелирует друг с другом. Слабее всего пара wine_servings и spirit_servings, но тоже положительная 0.194797

# дальше 2.10 я не делал

### 11

Каким образом следует осуществить соединение двух имеющихся у нас таблиц? Почему? Соедините таблицы!

Получилось ли у вас задействовать все строки, и если нет, то почему? Что делать с теми строками, которые не удается задействовать, если таковые есть? Ответ обоснуйте.

*Подсказка. В ответе на первый вопрос задания, а также в обосновании способа соединения таблиц вам может помочь задание 3 данного раздела ДЗ и соответствующий семинарский ноутбук!*

*Примечание. Не забудьте сохранить все изменения! В последующих заданиях вы должны работать с одним объединенным датасетом.*

In [67]:
# your code here

### 12

Осуществите группировку данных по странам и типу научной статьи (многомерная группировка). В качестве столбцов значений для группировки возьмите оцененную продолжительность жизни и общее потребление алкоголя в литрах. В качестве агреггирующих функций для продолжительности жизни возьмите среднее, медиану, минимум и максимум, а для потребления алкоголя — агреггирующую функцию/функции выберите самостоятельно. Как вы можете прокомментировать свой выбор и что вы наблюдаете?

In [68]:
# your code here

### 13

Вставьте любые два столбца из предыдущего задания, которые вы получили путем группировки, в исходную таблицу в качестве двух новых столбцов (назвать вы их можете как хотите). Разумеется, значения этих столбцов должны соответствовать значениям исходной таблицы по указанным в прошлом задании категориям.

*Примечание. Не забудьте сохранить все изменения! В последующих заданиях вы должны работать с обновленным датасетом.*

In [69]:
# your code here

### 14

Используя группировки, аргументированно ответьте на вопрос: на каких континентах средняя ожидаемая продолжительность жизни выше у мужчин, а на каких у женщин?

А медианная ожидаемая продолжительность?

А минимальная ожидаемая продолжительность?

Сделайте выводы!

*Подсказка. Казалось бы, при чем здесь тип научной статьи?*

In [70]:
# your code here

### 15

Работая лишь с общим объединенным датасетом и не используя исходную таблицу drinks, подсчитайте среднее потребление каждого вида алкоголя и среднее общее количество литров потребляемого алкоголя на человека по всем континентам. Занесите ваши результаты в отдельный датафрейм!

*Подсказка. Задача немного сложнее, чем может показаться на первый взгляд — в объединенном датасете страны встречаются разное количество раз, и из-за этого могут возникнуть определенные проблемы. Если так случится, надо подумать, как это обойти!*

*Примечание. Разумеется, это необходимо сделать автоматизированно, а не перебирая каждый континент по отдельности.*

In [71]:
# your code here

### 16

Создание новых признаков (feature engineering) является одним из основных средств улучшения качества работы алгоритмов машинного обучения на этапе обработки данных, а также отличный способ увеличить глубину ваших выводов из проделанного анализа в целом! Частично мы с вами уже осуществляли feature engineering, когда добавляли новые столбцы в задании 13 данного раздела.

Теперь давайте добавим в таблицу столбец, который будет показывать, какое общее количество servings per person производится в среднем по стране, к которой относится научная статья. Столбец для этого признака назовите 'total_servings'.

Чем может быть полезен такой признак?

***Бонус (2 бон. балла):*** Придумайте еще какой-нибудь свой осмысленный признак, который может быть полезен для анализа (несет в себе какую-то информацию), и добавьте его в таблицу. Обязательно аргументируйте, чем полезен ваш признак.

In [72]:
# your code here

### 17

Есть ли в данных категориальные признаки, которые принимают значения строк, а не чисел? Если да, то для многих алгоритмов это может стать серьезным препятствием, поэтому закодируйте категориальные признаки так, чтобы к данным можно было применить любые алгоритмы. Аргументируйте выбор методов кодирования.

*Подсказка. Вам могут помочь Label encoding и One-Hot Encoding. Но будьте внимательны! У каждого из этих методов есть свои сильные и слабые стороны, а также ограничения применимости.*

In [73]:
# your code here

### 18

Придумайте какое-нибудь своё интересное задание на анализ предлагаемых данных и решите его.

Креативность в этом задании приветствуется, и за нее можно получить ***1-2 бонусных балла***!

*Подсказка. Вы можете вдохновляться различными заданиями из этого раздела ДЗ, как, например, заданиями: 4, 6, 10, 12, 14, 15, 16. Но, чтобы получить бонусные баллы, лучше не используйте аналогичные идеи, а придумайте что-то своё!*

In [74]:
# your code here

## Бонусная часть на анализ целевой переменной (до 5 бонусных баллов)

После всего проделанного анализа и подготовки данных мы можем перейти непосредственно к ответу на исходный вопрос исследования, который был сформулирован вами в задании 3 данного раздела.

### 19

***Бонус (5 бон. баллов):*** Дайте развернутый, обоснованный, подкрепленный данными и написанным вами кодом ответ на вопрос, влияют ли признаки зависимой таблицы на целевую переменную из главной таблицы и, если да, то каким именно образом влияют?

Обязательно аргументируйте полученные вами результаты и выводы. Чем подробнее и обоснованнее анализ вы проведете в рамках данного бонусного задания, тем больше бонусных баллов получите! Вы можете использовать подходы и методы из рассмотренных на семинаре и в рамках данного домашнего задания, а можете придумать и что-то своё. Креативность максимально приветствуется!

In [75]:
# your code here

    {your final conclusion here}